### PR TITLE
Ben/new gui

### PR DESCRIPTION
**Pull request by Ben-Graham**

### CODE REVIEW

File: BusinessLayer/SongPlayer/SongPlayer.cpp [hunk 1]
```diff
@@ -185,11 +187,37 @@ void SongPlayer::toggleLoop()
...
 }
-
 void SongPlayer::updateInfo()
 {
     artist_ = mediaPlayer_.metaData(QMediaMetaData::ContributingArtist).toString();
     title_ = mediaPlayer_.metaData(QMediaMetaData::Title).toString();
-    cover_ = mediaPlayer_.metaData(QMediaMetaData::CoverArtImage).value<QImage>();
-    emit updateGUI(title_, artist_);
+    album_ = mediaPlayer_.metaData(QMediaMetaData::AlbumTitle).toString(); //retrieves the album name from current song
+    album_.replace(" ", ""); //remove all spaces in album name for easier access to file path of album
```
Comment (bill-luu) [inline BusinessLayer/SongPlayer/SongPlayer.cpp]: Add in some extra whitespace to separate ideas

File: BusinessLayer/SongPlayer/SongPlayer.cpp [hunk 2]
```diff
@@ -185,11 +187,37 @@ void SongPlayer::toggleLoop()
...
     }
 }
-
 void SongPlayer::updateInfo()
 {
     artist_ = mediaPlayer_.metaData(QMediaMetaData::ContributingArtist).toString();
     title_ = mediaPlayer_.metaData(QMediaMetaData::Title).toString();
-    cover_ = mediaPlayer_.metaData(QMediaMetaData::CoverArtImage).value<QImage>();
-    emit updateGUI(title_, artist_);
+    album_ = mediaPlayer_.metaData(QMediaMetaData::AlbumTitle).toString(); //retrieves the album name from current song
```
Comment (bill-luu) [inline BusinessLayer/SongPlayer/SongPlayer.cpp]: Don't need this comment

File: BusinessLayer/SongPlayer/SongPlayer.cpp (hunk 1 above)
Comment (bill-luu) [reply BusinessLayer/SongPlayer/SongPlayer.cpp]: Add the comment above this line

File: BusinessLayer/SongPlayer/SongPlayer.cpp [hunk 3]
```diff
@@ -185,11 +187,37 @@ void SongPlayer::toggleLoop()
...
+}
+
+QColor SongPlayer::getColor(QImage img)
+{
+    int height = img.height() / 2;
+    int width = img.width() / 2;
+    if(height != 0 && width != 0)
+    {
+        QColor color(img.pixel(width, height));
+        while(color.lightness() < 40)
```
Comment (bill-luu) [inline BusinessLayer/SongPlayer/SongPlayer.cpp]: `40` & `10` are pretty arbitrary magic numbers, so add them to a namespace

File: ViewLayer/ProgressBar/ProgressBar.cpp [hunk 4]
```diff
@@ -10,6 +10,7 @@ namespace
 
 ProgressBar::ProgressBar()
 {
+    color_ = QColor(92,145,161,255);
```
Comment (bill-luu) [inline ViewLayer/ProgressBar/ProgressBar.cpp]: Let's have the default color `QColor(92,145,161,255)` in the namespace

File: ViewLayer/SongPlayerView/SongPlayerView.cpp [hunk 5]
```diff
@@ -1,25 +1,26 @@
 #include "SongPlayerView.h"
 #include <QDebug>
```
Comment (bill-luu) [inline ViewLayer/SongPlayerView/SongPlayerView.cpp]: Remove `<QDebug>` from release code

File: ViewLayer/SongPlayerUI/SongPlayerUi.h [hunk 6]
```diff
@@ -20,13 +20,14 @@ class SongPlayerUi : public I_SongPlayerUi
...
+    QLabel& labelPic();
     QLabel& infoLabel();
-    QPushButton& PlayButton();
-    QProgressBar& ProgressBar();
-    QPushButton& NextSong();
-    QPushButton& PrevSong();
-    QPushButton& ShuffleButton();
-    QPushButton& LoopButton();
+    QPushButton& playButton();
+    QPushButton& shuffle();
```
Comment (bill-luu) [inline ViewLayer/SongPlayerUI/SongPlayerUi.h]: `shuffle` should be renamed to `shuffleButton`. Same with `loop`

File: ViewLayer/SongPlayerView/SongPlayerView.cpp [hunk 7]
```diff
@@ -73,9 +74,35 @@ void SongPlayerView::handleVolumeControl()
...
 {
     ui_.infoLabel().setText(artist + " - " + title);
+    if (!cover.isNull())
+    {
+        ui_.labelPic().setPixmap(cover);
+        ui_.labelPic().setScaledContents(true);
+    }
+    QImage img(cover.toImage());
+    QColor color = songPlayer_.getColor(img);
+    QString styleSheet = "QSlider::groove:vertical {"
```
Comment (bill-luu) [inline ViewLayer/SongPlayerView/SongPlayerView.cpp]: This stylesheet can go in the namespace

File: BusinessLayer/SongPlayer/SongPlayer.cpp [hunk 8]
```diff
@@ -185,11 +187,37 @@ void SongPlayer::toggleLoop()
...
+    cover_.replace(cover_.lastIndexOf(ALBUM_FILE_PATH) + ALBUM_FILE_PATH.length(), songNameLength, album_);
+    QPixmap img(cover_);
+    emit updateGUI(title_, artist_, img);
+}
+
+QColor SongPlayer::getColor(QImage img)
+{
+    int height = img.height() / 2;
+    int width = img.width() / 2;
+    if(height != 0 && width != 0)
```
Comment (bill-luu) [inline BusinessLayer/SongPlayer/SongPlayer.cpp]: Add a comment to explain exactly when this isn't true.

File: DataLayer/SongLibrary/Albumcheck.sh [hunk 9]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+
+song=(1-06+Animal 1-11+Armageddon+It 1-14+Hysteria 01-Take-On-Me)
```
Comment (bill-luu) [inline DataLayer/SongLibrary/Albumcheck.sh]: I think we agreed to delete this

**Review by bill-luu — CHANGES_REQUESTED** [review]

**Comment by bill-luu** [comment]
Right now, if the song has no cover the volume slider's color bar turns black:
image

**Comment by ExtraGravity** [comment]
Nice icons!

File: ViewLayer/SongPlayerView/SongPlayerView.cpp [hunk 10]
```diff
@@ -19,6 +19,8 @@ namespace
 
                               "QSlider::sub-page:vertical {"
                               "background: white;}";
+
+    const QColor DEFAULT_COLOR = QColor(255,255,255,255);
```
Comment (bill-luu) [inline ViewLayer/SongPlayerView/SongPlayerView.cpp]: Match this color with the default color of the other icons

**Review by bill-luu — CHANGES_REQUESTED** [review]
One more thing, and then I'll merge this.

**Review by bill-luu — APPROVED** [review]

**Pull request merged**